design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+  objects, starting package controllers, and cleaning up these objects when a
+  package is upgraded.
+
+These types and controllers overlap in functionality with some of the existing
+types and controllers that are currently present in the package manager
+ecosystem. However, their implementation should not immediately break any of the
+current features supported by the package manager. Removal of
+`ClusterPackageInstall`, `PackageInstall`, and `StackDefinition`, as well as
+refactor of `Package` and all controllers should follow a deprecation schedule
+that allows users to move to the new format.
+
+## Implementation
+
+This following sections describe the changes to the current package manager that
+are required to support the goals listed above.
+
+- [Current Workflow](#current-workflow)
+- [Future Workflow](#future-workflow)
+- [Phase 1](#phase-1)
+- [Phase 2](#phase-2)
+- [Phase 3](#phase-3)
+
+### Current Workflow
+
+In order to accurately describe the steps required for implementing this
+refactor. It is important to understand, at least at a high-level, the current
+workflow the package manager uses for installing a `Package`. We will use a
+`ClusterPackageInstall` for the example here as the namespace-scoped
+`PackageInstall` is intended to be removed as part of this refactor.
+
+1. User creates a `ClusterPackageInstall`.
+2. The _`ClusterPackageInstall` controller_ observes the `ClusterPackageInstall`
+   creation and creates a `Job` that runs a `Pod` with an initContainer running
+   the image supplied on the `ClusterPackageInstall` and a container running the
+   same image the package manager itself is running. The `initContainer` serves
+   to only copy the contents of the package directory from the supplied image
+   into a `Volume`. That `Volume` is then mounted on the `container` running the
+   package manager image, which is executed with the `unpack`
+   [argument](https://github.com/crossplane/crossplane/blob/a0d139f7cf269599ba916ed15af3fd68ffeabbdf/cmd/crossplane/package/unpack/unpack.go#L52).
+3. The `unpack` container walks the filepath of the package directory, using the
+   information provided to construct a `Package` object. It then
+   [writes](https://github.com/crossplane/crossplane/blob/a0d139f7cf269599ba916ed15af3fd68ffeabbdf/pkg/packages/unpack.go#L204)
+   the `Package` object and all CRDs in the package to
+   [stdout](https://github.com/crossplane/crossplane/blob/a0d139f7cf269599ba916ed15af3fd68ffeabbdf/cmd/crossplane/package/unpack/unpack.go#L53).
+4. The _`ClusterPackageInstall` controller_ waits for the `Job` to complete
+   successfully before reading the logs from the `Pod`. When the `Job` is
+   complete, it reads the the logs and creates all of the objects that were
+   printed, making a [few modifications as well as annotating and labelling
+   appropriately](https://github.com/crossplane/crossplane/blob/6fc50822fbf11a7d31f8a9dabde5c8948c3b36ac/pkg/controller/packages/install/installjob.go#L259).
```

Comment:
Basically what we are looking at here is unpacking a package locally and then applying its contents, which is technically possible in this proposal, but would end up being somewhat redundant (the user could create the `PackageRevision` from the output and then update the `Package` to use that image, but the `Package` controller would still need to create the `Job` to get the hash of the image, so it would be somewhat redundant). As far as other options, I have explored a few (whichever we decide on I will make sure to update this doc with reasoning):

- Connecting to the host container runtime directly from the package manager `Pod`. This would likely be the cleanest but would also introduce some additional complexity. You essentially end up managing containers alongside Kubernetes on the same host, which makes for poor separation of concern and can cause issues with resource consumption, etc. Functionally, the package manager would pull the image directly on to the host, then analyze the package contents by parsing the image layers.
- **[Most Viable IMO]** Running docker in docker (i.e. Docker daemon in the pod, likely in a sidecar container). This is a bit more secure, but also has some of its own complexities. The benefit here is that these operations are isolated from the actual host container runtime, and the package manager pod and the DinD pod would just communicate over localhost. We are not interested in actually running containers, only inspecting their contents here, so we don't really have to worry too much about managing the stability of the DinD process (we can just kill it and recreate whenever needed).
  - Side note: I explored trying to pull image layers in-memory, but this isn't really possible in a generic way from any container registry.
- Using an `unpack` image as the base image. This was actually explored in the [original package design](https://github.com/crossplane/crossplane/blob/master/design/design-doc-packages.md#alternate-packaging-designs). It appears this would drastically increase the size of the container image, so I mostly shied away from this. Would be interested in @jbw976 has any more context?
- Package container pipe to `kubectl`. If we want to be able to make decisions about whether it is safe to upgrade a package to a new version I think this model is difficult. I guess we would be building the `unpack` logic in as with the previous case, so we could put logic in there, but I am thinking it would be quite messy.

Reply:
> This was actually explored in the original package design. It appears this would drastically increase the size of the container image, so I mostly shied away from this.

I wonder how impactful this could be? I could imagine it being fairly small-ish if it was a separate build/binary from core Crossplane. Plus I imagine it would not be stored/pulled with every package given OCI layering.

Reply:
Also just to be clear, given that the whole unpack logic is fairly behind the scenes I could be convinced that we should punt on it for the time being and focus on user-facing API changes. It seems like we could probably change this even once the API hits v1beta1 without breaking changes.

Reply:
IMO, docker-in-docker solutions either pose security risks by accessing host container runtime or end up introducing complexity, as mentioned in @hasheddan 's comment.

> Package container pipe to kubectl. If we want to be able to make decisions about whether it is safe to upgrade a package to a new version I think this model is difficult. I guess we would be building the unpack logic in as with the previous case, so we could put logic in there, but I am thinking it would be quite messy.

I actually think that's a good solution. The whole processing would be **moved over** to `unpack` command and in the output of the job we'd only see `kubectl`'s output. I remember a case where in OpenShift, the log rotation was enabled and long `unpack` jobs like the ones from providers were cut off quite early. So, relying on the stdout sounds risky to me.

Reply:
@muvaf @negz when you say "pipe to kubectl" I am assuming you mean creating the objects with `client-go` right? Or is there some advantage in actually piping them to kubectl?

Reply:
I think I'm saying `pipe to kubectl` because I'm used to see the YAML output in the stdout but yeah just creating them via existing controller runtime would work better.

Reply:
Going to resolve this comment thread because the format of packaging and unpacking has been moved out of the scope of this design doc :+1: I will make sure these thoughts are reflected in the `Related Efforts` section